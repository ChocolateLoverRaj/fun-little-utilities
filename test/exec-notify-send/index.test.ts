import fsx from 'fs-extra'
import temp from 'unique-temp-path'

import defaultExport, {
  notifySend,
  ExecFunc,
  Urgency,
  HintType
} from 'exec-notify-send'

const mockedExecute = (fn: ExecFunc = () => ({})) => jest.fn(fn)

it('export default notifySend', () => {
  expect(defaultExport).toBe(notifySend)
})

it('default command is "notify-send"', async () => {
  expect.assertions(1)

  const execute: ExecFunc = command => {
    expect(command).toBe('notify-send')
    return {}
  }

  await notifySend({
    summary: 'summary',
    execute
  })
})

it('default execute function works', async () => {
  const script = temp('test-exec-notify-send.script.', '.js')
  const output = temp('test-exec-notify-send.output.', '.txt')

  const scriptContent = `
    #! /usr/bin/env node
    const process = require('process')
    const fs = require('fs')
    const text = process.argv.slice(2).join(' ')
    fs.writeFileSync(${JSON.stringify(output)}, text)
    process.exit(0)
  `
    .split(/\n|\r/)
    .map(line => line.trim())
    .filter(Boolean)
    .join('\n')

  await fsx.writeFile(script, scriptContent)
  await fsx.chmod(script, '755')

  await notifySend({
    command: script,
    summary: 'summary',
    body: 'body',
    appName: 'app-name',
    icon: ['32x', '16x', '8x'],
    expireTime: 123
  })

  expect(await fsx.readFile(output, 'utf8')).toMatchSnapshot()
})

it('summary only', async () => {
  const execute = mockedExecute()
  await notifySend({
    summary: 'summary',
    execute
  })
  expect(execute.mock.calls).toMatchSnapshot()
})

it('summary and body', async () => {
  const execute = mockedExecute()
  await notifySend({
    summary: 'summary',
    body: 'body',
    execute
  })
  expect(execute.mock.calls).toMatchSnapshot()
})

it('more options', async () => {
  const execute = mockedExecute()
  await notifySend({
    summary: 'summary',
    body: 'body',
    appName: 'Application Name',
    icon: ['x64', 'x32', 'x16', 'x8'],
    category: ['test', 'category'],
    expireTime: 123,
    urgency: Urgency.Critical,
    hint: { name: 'hint', type: HintType.Int, value: 35 },
    execute
  })
  expect(execute.mock.calls).toMatchSnapshot()
})
