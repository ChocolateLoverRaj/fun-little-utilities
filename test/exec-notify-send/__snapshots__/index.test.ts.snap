// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`default execute function works 1`] = `"--app-name=app-name --icon=32x,16x,8x --expire-time=123 -- summary body"`;

exports[`more options 1`] = `
Array [
  Array [
    "notify-send",
    Array [
      "--app-name=Application Name",
      "--icon=x64,x32,x16,x8",
      "--urgency=critical",
      "--expire-time=123",
      "--category=test,category",
      "--hint=int:hint:35",
      "--",
      "summary",
      "body",
    ],
  ],
]
`;

exports[`summary and body 1`] = `
Array [
  Array [
    "notify-send",
    Array [
      "--",
      "summary",
      "body",
    ],
  ],
]
`;

exports[`summary only 1`] = `
Array [
  Array [
    "notify-send",
    Array [
      "--",
      "summary",
    ],
  ],
]
`;
